// @sap/approuter/lib/backend-request/agents.js
'use strict';

const Cache = require('lru-cache');
const { HttpProxyAgent } = require('http-proxy-agent');
const { HttpsProxyAgent } = require('https-proxy-agent');
const { Agent: HttpAgent } = require('http');
const { Agent: HttpsAgent } = require('https');
const HttpKeepAliveAgent = require('agentkeepalive');
const HttpsKeepAliveAgent = HttpKeepAliveAgent.HttpsAgent;
const isConReuseDisabled = isConnectionReuseDisabled();

function isRunningOnCF() {
  try {
    const vcapApp = JSON.parse(process.env.VCAP_APPLICATION || '{}');
    return !!vcapApp.cf_api;
  } catch (e) {
    return false;
  }
}

function isConnectionReuseDisabled() {
  const disableValue = process.env.DISABLE_CONNECTION_REUSE;
  if (disableValue === 'true') {
    return true;
  } else if (disableValue === 'false') {
    return false;
  } else if (!disableValue && !isRunningOnCF()) {
    return true;
  }
  return false;
}

if (isConReuseDisabled) {
  exports.proxyAgentsCache = new Cache(1);
} else {
  exports.proxyAgentsCache = new Cache(10);
}
exports.httpAgent = new HttpKeepAliveAgent();
exports.httpsAgent = new HttpsKeepAliveAgent();


function isSecureConnection(protocol) {
  return protocol === 'https:' || protocol === 'wss:';
}

exports.get = function (protocol, proxy, connectivityScenario) {
  let isSecure = connectivityScenario ? false : isSecureConnection(protocol);
  if (proxy) {
    return retrieveProxyAgent(proxy, isSecure);
  }
  if (isConReuseDisabled) {
    return isSecure ? new HttpsAgent() : new HttpAgent();
  }
  return isSecure ? exports.httpsAgent : exports.httpAgent;
};

function retrieveProxyAgent(proxy, isSecure) {
  let id = buildIdentifier(proxy, isSecure);
  let agent = exports.proxyAgentsCache.get(id);
  if (agent) {
    return agent;
  }
  agent = isSecure ? new HttpsProxyAgent(proxy) : new HttpProxyAgent(proxy);
  exports.proxyAgentsCache.set(id, agent);
  return agent;
}

function buildIdentifier(proxy, isSecure) {
  let suffix = isSecure ? 'secure' : 'plain';
  return proxy + ':' + suffix;
}
